File: _reference/backend/src/features/notification/service.ts
import db from '@/config/database';
import { notifications as notificationsTable } from '@/models/notifications';
import { eq, and, sql, desc, type SQL } from 'drizzle-orm';
import { NotFoundError } from '@/config/error';
import { getPagination } from '@/db/utils/query';
import type { Notification } from '@repo/types';

export class NotificationService {
    /**
     * 获取通知列表
     */
    static async getNotifications({ page, limit, unreadOnly = 'false' }: { page?: number; limit?: number; unreadOnly?: string }): Promise<{
        data: Notification[];
        page: number;
        limit: number;
        total: number;
    }> {
        const filters: SQL[] = [];

        const { pageNum, limitNum, offset } = getPagination(page, limit);

        if (unreadOnly === 'true') {
            filters.push(eq(notificationsTable.isRead, false));
        }

        const where = filters.length > 0 ? and(...filters) : undefined;

        const [notifications, total] = await Promise.all([
            db.query.notifications.findMany({
                where,
                orderBy: [desc(notificationsTable.createdAt)],
                limit: limitNum,
                offset
            }),
            db
                .select({ count: sql<number>`count(*)` })
                .from(notificationsTable)
                .where(where)
        ]);

        const data = notifications.map(item => ({
            ...item,
            createdAt: item.createdAt.toISOString() || ''
        }));

        return {
            data,
            page: pageNum,
            limit: limitNum,
            total: total[0]?.count || 0
        };
    }

    /**
     * 标记为已读
     */
    static async markAsRead(id: string) {
        const result = await db.update(notificationsTable).set({ isRead: true }).where(eq(notificationsTable.id, id)).returning();

        if (result.length === 0) {
            throw new NotFoundError('通知未找到');
        }

        return result[0];
    }

    /**
     * 全部标记为已读
     */
    static async markAllAsRead(): Promise<Notification[]> {
        const result = await db.update(notificationsTable).set({ isRead: true }).where(eq(notificationsTable.isRead, false)).returning();

        return result.map(item => ({
            ...item,
            createdAt: item.createdAt.toISOString() || ''
        }));
    }

    /**
     * 删除通知
     */
    static async delete(id: string) {
        const result = await db.delete(notificationsTable).where(eq(notificationsTable.id, id)).returning();

        if (result.length === 0) {
            throw new NotFoundError('通知未找到');
        }

        return true;
    }

    /**
     * 清空所有通知
     */
    static async clearAll(): Promise<Notification[]> {
        const result = await db.delete(notificationsTable).returning();

        return result.map(item => ({
            ...item,
            createdAt: item.createdAt.toISOString() || ''
        }));
    }
}

File: internal/notification/service.go
package notification

import (
	"context"
	"time"

	"wood-hot-monitor/ent"
	entnotif "wood-hot-monitor/ent/notification"
	"wood-hot-monitor/internal/database"
	"wood-hot-monitor/internal/models"
)

type Service struct {
	db *database.DB
}

func NewService(db *database.DB) *Service {
	return &Service{db: db}
}

type GetAllParams struct {
	Page       int  `json:"page"`
	Limit      int  `json:"limit"`
	UnreadOnly bool `json:"unreadOnly"`
}

func (s *Service) GetAll(params GetAllParams) (*models.PaginatedResult[models.Notification], error) {
	ctx := context.Background()

	query := s.db.Client.Notification.Query()
	if params.UnreadOnly {
		query = query.Where(entnotif.IsRead(false))
	}

	total, err := query.Clone().Count(ctx)
	if err != nil {
		return nil, err
	}

	offset := (params.Page - 1) * params.Limit
	rows, err := query.
		Order(ent.Desc(entnotif.FieldCreatedAt)).
		Limit(params.Limit).
		Offset(offset).
		All(ctx)
	if err != nil {
		return nil, err
	}

	data := make([]models.Notification, len(rows))
	for i, row := range rows {
		data[i] = mapNotification(row)
	}

	return &models.PaginatedResult[models.Notification]{
		Data:  data,
		Total: total,
		Page:  params.Page,
		Limit: params.Limit,
	}, nil
}

func (s *Service) MarkAsRead(id string) error {
	return s.db.Client.Notification.UpdateOneID(id).
		SetIsRead(true).
		Exec(context.Background())
}

func (s *Service) MarkAllAsRead() error {
	_, err := s.db.Client.Notification.Update().
		Where(entnotif.IsRead(false)).
		SetIsRead(true).
		Save(context.Background())
	return err
}

func (s *Service) Delete(id string) error {
	return s.db.Client.Notification.DeleteOneID(id).Exec(context.Background())
}

func (s *Service) ClearAll() error {
	_, err := s.db.Client.Notification.Delete().Exec(context.Background())
	return err
}

func (s *Service) Create(typ, title, content string, hotspotID *string) (*models.Notification, error) {
	builder := s.db.Client.Notification.Create().
		SetType(typ).
		SetTitle(title).
		SetContent(content).
		SetIsRead(false).
		SetCreatedAt(time.Now().UTC())
	if hotspotID != nil {
		builder = builder.SetHotspotID(*hotspotID)
	}

	row, err := builder.Save(context.Background())
	if err != nil {
		return nil, err
	}

	m := mapNotification(row)
	return &m, nil
}

func mapNotification(row *ent.Notification) models.Notification {
	return models.Notification{
		ID:        row.ID,
		Type:      row.Type,
		Title:     row.Title,
		Content:   row.Content,
		IsRead:    row.IsRead,
		HotSpotID: row.HotspotID,
		CreatedAt: row.CreatedAt.Format(time.DateTime),
	}
}
